### PR TITLE
PADV-806: feat: add new content authoring event signals

### PR DESCRIPTION
**Pull request by JuanDavidBuitrago**
## Description

Use the following ``OpenEdxPublicSignal`` events triggered by content creation and editing.

* XBLOCK_CREATED and XBLOCK_UPDATED
* COURSE_CREATED
* CONTENT_LIBRARY_CREATED, CONTENT_LIBRARY_UPDATED and CONTENT_LIBRARY_DELETED
* LIBRARY_BLOCK_CREATED, LIBRARY_BLOCK_UPDATED and LIBRARY_BLOCK_DELETED

## Testing instruction

1. Use this branch ``vue/PADV-806`` to use the necessary changes
2. Setup frontend-app-library-authoring if not done before:

```bash
    from .devstack import FEATURES
    FEATURES['ENABLE_LIBRARY_AUTHORING_MICROFRONTEND'] = True
    BLOCKSTORE_API_AUTH_TOKEN = 'edxapp-insecure-devstack-key'
```
3. Set up the blockstore to run in devstack ([ref](https://github.com/openedx/blockstore)) by running the following from the studio shell (make studio-shell):
```bash
# Configure blockstore to run inside LMS/Studio (instead of as an external service)
./manage.py cms waffle_switch --create blockstore.use_blockstore_app_api on

./manage.py cms shell  # enter python code below

# Create a "Collection" that new content libraries / xblocks can be created within:
from blockstore.apps.bundles.models import Collection
coll, _ = Collection.objects.get_or_create(title='Devstack Content Collection', uuid='11111111-2111-4111-8111-111111111111')

# Create an "Organization":
from organizations.models import Organization
Organization.objects.get_or_create(short_name='DeveloperInc', defaults={'name': 'DeveloperInc', 'active': True})
```
4. Add below snippet somewhere, for example in cms/djangoapps/contentstore/signals/handlers.py:

```bash
from openedx_events.content_authoring.signals import (
    XBLOCK_CREATED,
    XBLOCK_UPDATED,
    COURSE_CREATED,
    CONTENT_LIBRARY_CREATED,
    CONTENT_LIBRARY_UPDATED,
    CONTENT_LIBRARY_DELETED,
    LIBRARY_BLOCK_CREATED,
    LIBRARY_BLOCK_UPDATED,
    LIBRARY_BLOCK_DELETED,
)
XBLOCK_CREATED.connect(lambda **x: print("test_event: ", x))
XBLOCK_UPDATED.connect(lambda **x: print("test_event: ", x))
COURSE_CREATED.connect(lambda **x: print("test_event: ", x))
CONTENT_LIBRARY_CREATED.connect(lambda **x: print("test_event: ", x))
CONTENT_LIBRARY_UPDATED.connect(lambda **x: print("test_event: ", x))
CONTENT_LIBRARY_DELETED.connect(lambda **x: print("test_event: ", x))
LIBRARY_BLOCK_CREATED.connect(lambda **x: print("test_event: ", x))
LIBRARY_BLOCK_UPDATED.connect(lambda **x: print("test_event: ", x))
LIBRARY_BLOCK_DELETED.connect(lambda **x: print("test_event: ", x))
```

5. Open studio in http://localhost:18010/
6. Open studio logs monitoring:
```bash
make studio-logs | grep test_event:
```
7. Create a course
Check the log for ``COURSE_CREATED`` and the XBLOCK_UPDATED event data

8. Should see something like:

```bash
make studio-logs | grep test_event:
edx.devstack-olive.master.studio   | test_event:  {'signal': <OpenEdxPublicSignal: org.openedx.content_authoring.xblock.updated.v1>, 'sender': None, 'xblock_info': XBlockData(usage_key=BlockUsageLocator(CourseLocator('demo', 'demo1', '2020', None, None), 'course', 'course'), block_type='course', version=BlockUsageLocator(CourseLocator('demo', 'demo1', '2020', 'draft-branch', ObjectId('655df4e81593fe8f4610c9a8')), 'course', 'course')), 'metadata': EventsMetadata(event_type='org.openedx.content_authoring.xblock.updated.v1', id=UUID('39d5e9d2-8933-11ee-8a88-0242ac130010'), minorversion=0, source='openedx/cms/web', sourcehost='studio.devstack.edx', time=datetime.datetime(2023, 11, 22, 12, 32, 40, 172109, tzinfo=datetime.timezone.utc), sourcelib=(8, 3, 0))}
edx.devstack-olive.master.studio   | test_event:  {'signal': <OpenEdxPublicSignal: org.openedx.content_authoring.xblock.created.v1>, 'sender': None, 'xblock_info': XBlockData(usage_key=BlockUsageLocator(CourseLocator('demo', 'demo2', '2020', None, None), 'course_info', 'updates'), block_type='course_info', version=BlockUsageLocator(CourseLocator('demo', 'demo2', '2020', 'draft-branch', ObjectId('655e011b89ca06c615ebfe56')), 'course_info', 'updates')), 'metadata': EventsMetadata(event_type='org.openedx.content_authoring.xblock.created.v1', id=UUID('7fb4138c-893a-11ee-a5a4-0242ac130010'), minorversion=0, source='openedx/cms/web', sourcehost='studio.devstack.edx', time=datetime.datetime(2023, 11, 22, 13, 24, 43, 867932, tzinfo=datetime.timezone.utc), sourcelib=(8, 3, 0))}
```

## Jira issue

* https://agile-jira.pearson.com/browse/PADV-806

## Other information

Reference PR edx-platform: https://github.com/openedx/edx-platform/pull/32599
Reference PR openedx/events: https://github.com/openedx/openedx-events/pull/244

### CODE REVIEW

**Review by Jacatove — COMMENTED** [review]
@JuanDavidBuitrago Do we really need to back port all content authoring event signals?.
For me it is clear the need to do it for COURSE_CREATED, but I don't know about the others.

File: openedx/core/djangoapps/content_libraries/api.py [hunk 1]
```diff
@@ -802,7 +820,12 @@ def create_library_block(library_key, block_type, definition_id):
     # Clear the bundle cache so everyone sees the new block immediately:
     BundleCache(ref.bundle_uuid, draft_name=DRAFT_NAME).clear()
     # Now return the metadata about the new block:
-    LIBRARY_BLOCK_CREATED.send(sender=None, library_key=ref.library_key, usage_key=usage_key)
+    LIBRARY_BLOCK_CREATED.send_event(
+        library_block=LibraryBlockData(
+            library_key=ref.library_key,
+            usage_key=usage_key
```
Comment (anfbermudezme) [inline openedx/core/djangoapps/content_libraries/api.py:826]: Trailing comma. Review the other modules.

File: xmodule/modulestore/tests/test_mixed_modulestore.py [hunk 2]
```diff
@@ -724,6 +739,134 @@ def test_publish_automatically_after_delete_unit(self, default_ms):
...
+        assert event_receiver.call_args.kwargs['xblock_info'].block_type == sequential.location.block_type
+        assert event_receiver.call_args.kwargs['xblock_info'].version.for_branch(None) == sequential.location
+
+    @ddt.data(ModuleStoreEnum.Type.mongo, ModuleStoreEnum.Type.split)
+    def test_xblock_publish_event(self, default_ms):
+        """
+        Check that XBLOCK_PUBLISHED event is sent when xblock is published.
+        """
+        self.initdb(default_ms)
+        event_receiver = Mock()
```
Comment (anfbermudezme) [inline xmodule/modulestore/tests/test_mixed_modulestore.py:821]: Add blank line.

File: xmodule/modulestore/tests/test_mixed_modulestore.py [hunk 3]
```diff
@@ -724,6 +739,134 @@ def test_publish_automatically_after_delete_unit(self, default_ms):
...
+        XBLOCK_PUBLISHED.connect(event_receiver)
+        test_course = self.store.create_course('test_org', 'test_course', 'test_run', self.user_id)
+        # create sequential and vertical to test against
+        sequential = self.store.create_child(self.user_id, test_course.location, 'sequential', 'test_sequential')
+        self.store.create_child(self.user_id, sequential.location, 'vertical', 'test_vertical')
+        # publish sequential changes
+        self.store.publish(sequential.location, self.user_id)
```
Comment (anfbermudezme) [inline xmodule/modulestore/tests/test_mixed_modulestore.py:828]: Add blank line. We follow the structure proposed in https://3.basecamp.com/3966315/buckets/12732851/documents/2737536654. As a summary we divide the tests in three blocks. mock things, call things and assert things.

File: xmodule/modulestore/tests/test_mixed_modulestore.py [hunk 4]
```diff
@@ -15,6 +15,14 @@
 from unittest.mock import Mock, call, patch
 
 import ddt
+from openedx_events.content_authoring.data import CourseData, XBlockData
+from openedx_events.content_authoring.signals import (
+    COURSE_CREATED,
+    XBLOCK_CREATED,
+    XBLOCK_DELETED,
+    XBLOCK_PUBLISHED,
+    XBLOCK_UPDATED
```
Comment (anfbermudezme) [inline xmodule/modulestore/tests/test_mixed_modulestore.py:24]: Trailing comma.

File: xmodule/modulestore/mixed.py [hunk 5]
```diff
@@ -911,7 +970,16 @@ def publish(self, location, user_id, **kwargs):
         Returns the newly published item.
         """
         store = self._verify_modulestore_support(location.course_key, 'publish')
-        return store.publish(location, user_id, **kwargs)
+        item = store.publish(location, user_id, **kwargs)
+        # .. event_implemented_name: XBLOCK_PUBLISHED
```
Comment (anfbermudezme) [inline xmodule/modulestore/mixed.py:974]: Why do you think this inline commets are necessary?

File: xmodule/modulestore/mixed.py [hunk 6]
```diff
@@ -13,6 +13,16 @@
 from opaque_keys.edx.keys import CourseKey
 from opaque_keys.edx.locator import LibraryLocator
 
+from openedx_events.content_authoring.data import CourseData, XBlockData
+from openedx_events.content_authoring.signals import (
+    COURSE_CREATED,
+    XBLOCK_CREATED,
+    XBLOCK_DELETED,
+    XBLOCK_PUBLISHED,
+    XBLOCK_UPDATED
```
Comment (anfbermudezme) [inline xmodule/modulestore/mixed.py:22]: Trailing comma.

File: xmodule/modulestore/tests/test_mixed_modulestore.py [hunk 7]
```diff
@@ -481,7 +496,7 @@ def test_get_items_include_orphans(self, default_ms, expected_items_in_tree, orp
         assert {'course', 'about'}.issubset({item.location.block_type for item in items})  # pylint: disable=line-too-long
```
Comment (anfbermudezme) [inline xmodule/modulestore/tests/test_mixed_modulestore.py:496]:
```suggestion
        assert {'course', 'about'}.issubset(
           {item.location.block_type for item in items}
        )
```

File: xmodule/modulestore/tests/test_mixed_modulestore.py [hunk 8]
```diff
@@ -481,7 +496,7 @@ def test_get_items_include_orphans(self, default_ms, expected_items_in_tree, orp
...
         # Assert that about is a detached category found in get_items
         assert [item.location.block_type for item in items if item.location.block_type == 'about'][0]\
```
Comment (anfbermudezme) [inline xmodule/modulestore/tests/test_mixed_modulestore.py:498]:
```suggestion
        assert [
            item.location.block_type for item in items if item.location.block_type == 'about'
        ][0] in **DETACHED_XBLOCK_TYPES**
```

File: openedx/core/djangoapps/content_libraries/tests/test_content_libraries.py [hunk 9]
```diff
@@ -491,7 +502,7 @@ def test_library_blocks_filters(self, is_indexing_enabled):
             assert len(self._get_library_blocks(lib['id'], {'text_search': 'Foo', 'block_type': 'video'})) == 0
             assert len(self._get_library_blocks(lib['id'], {'text_search': 'Baz', 'block_type': 'video'})) == 1
             assert len(self._get_library_blocks(lib['id'], {'text_search': 'Baz', 'block_type': ['video', 'html']})) ==\
```
Comment (anfbermudezme) [inline openedx/core/djangoapps/content_libraries/tests/test_content_libraries.py:504]: Review indentation please.

**Review by anfbermudezme — CHANGES_REQUESTED** [review]

**Comment by JuanDavidBuitrago** [comment]
> @JuanDavidBuitrago Do we really need to back port all content authoring event signals?. For me it is clear the need to do it for COURSE_CREATED, but I don't know about the others.

I added the complete content authoring event signals because I had to update the [openedx-events](https://github.com/openedx/openedx-events/pull/244) [version](https://github.com/Pearson-Advance/edx-platform/pull/108/files#diff-160eb30f60d519b8fe8a919cb2646a7ebec7652952129cfdd803776e70be6300R762), so I consider it is better to implement all of them to avoid problem using that version.

**Comment by JuanDavidBuitrago** [comment]
I will let it as a draft to future work if it's useful.

Here is the PR with COURSE_CREATED signal only: https://github.com/Pearson-Advance/edx-platform/pull/109

**Pull request closed without merging**